### ISSUE TITLE
az repos policy merge-strategy update failing

### ISSUE THREAD

**Issue by gauravsaralMs** [Azure/azure-devops-cli-extension#858]
Hi,

I wanted to report a bug for the azure-devops module of the az cli.

If you try to disable a merge policy for a branch, it fails with his error:
az repos policy merge-strategy update --id 80581 --enabled false
TF402457: The settings for this policy are not correctly formatted. Error: Error converting value {null} to type 'System.Boolean'. Path 'useSquashMerge', line 9, position 24. 

I think this is because ADO changed their policy format recently, but azure-devops is still using the old useSquashMerge instead of allowSquash.

Thanks,
David

**Comment by davidtorosyan** [comment]
I sent the original bug report, here's a more detailed repro as well as debug logs.

First, a note about the version. I first repro'd this on azure-devops extension version 0.10.0, where it fails with the error above. I've since updated to 0.13.0, and now it repros with this error:
`TF402457: The settings for this policy are not correctly formatted. Error: Error converting value {null} to type 'System.Boolean'. Path 'allowRebase', line 10, position 21.`

## Repro info

### Steps

1. Start with an ADO branch with no merge policies configured:
![image](https://user-images.githubusercontent.com/46736285/67970485-4ddef880-fbc8-11e9-8b33-0e812ee6fb13.png)
2. Configure one option, doesn't matter which one:
![image](https://user-images.githubusercontent.com/46736285/67972058-2a697d00-fbcb-11e9-801d-7eee96ed1281.png)
3. Get the policy:
```powershell
PS> az repos policy list -o json --repository-id <repoId> --branch <branch> --query "[?type.id=='fa4e907d-c16b-4a4c-9dfa-4916e5d171ab'].{ID:id,IsBlocking:isBlocking,IsEnabled:isEnabled,Name:type.displayName,DisplayName:settings.displayName,Type:type.id,Settings:settings}"
[
  {
    "DisplayName": null,
    "ID": <policyId>,
    "IsBlocking": true,
    "IsEnabled": false,
    "Name": "Require a merge strategy",
    "Settings": {
      "allowSquash": true,
      "scope": [
        {
          "matchKind": "Exact",
          "refName": "refs/heads/<branch>",
          "repositoryId": "<repoId>"
        }
      ]
    },
    "Type": "fa4e907d-c16b-4a4c-9dfa-4916e5d171ab"
  }
]
```

4. Try to disable the policy, and it fails:
```powershell
> az repos policy merge-strategy update --id <policyId> --enabled false
TF402457: The settings for this policy are not correctly formatted. Error: Error converting value {null} to type 'System.Boolean'. Path 'allowRebase', line 10, position 21.
```

5. However, if you specify any of the options, it works fine:
```powershell
> az repos policy merge-strategy update --id <policyId> --enabled false --allow-squash true
# success!
```

### Debug logs

Running step (4) with `--debug`, we see the problem (it's sending None instead of false):
```powershell
azext_devops.devops_sdk.client : Request content: {'type': {'id': 'fa4e907d-c16b-4a4c-9dfa-4916e5d171ab'}, 'isBlocking': True, 'isEnabled': False, 'settings': {'scope': [{'repositoryId': '<repoId>', 'refName': 'refs/heads/<branch>', 'matchKind': 'Exact'}], 'allowSquash': True, 'allowRebase': None, 'allowRebaseMerge': None, 'allowNoFastForward': None}}
msrest.universal_http : Configuring redirects: allow=True, max=30
msrest.universal_http : Configuring request: timeout=100, verify=True, cert=None
msrest.universal_http : Configuring proxies: ''
msrest.universal_http : Evaluate proxies against ENV settings: True
urllib3.connectionpool : https://<project>.visualstudio.com:443 "PUT /<projectId>/_apis/policy/configurations/<policyId> HTTP/1.1" 400 411
azext_devops.devops_sdk.client : Response content: b'{"$id":"1","innerException":null,"message":"TF402457: The settings for this policy are not correctly formatted. Error: Error converting value {null} to type \'System.Boolean\'. Path \'allowRebase\', line 10, position 21.","typeName":"Microsoft.TeamFoundation.Policy.Server.PolicySettingsFormatException, Microsoft.TeamFoundation.Policy.Server","typeKey":"PolicySettingsFormatException","errorCode":0,"eventId":3000}'
msrest.exceptions : TF402457: The settings for this policy are not correctly formatted. Error: Error converting value {null} to type 'System.Boolean'. Path 'allowRebase', line 10, position 21.
```

And we can see how step (5) is different:
```powershell
azext_devops.devops_sdk.client : Request content: {'type': {'id': 'fa4e907d-c16b-4a4c-9dfa-4916e5d171ab'}, 'isBlocking': True, 'isEnabled': False, 'settings': {'scope': [{'repositoryId': '<repoId>', 'refName': 'refs/heads/<branch>', 'matchKind': 'Exact'}], 'allowSquash': True, 'allowRebase': False, 'allowRebaseMerge': False, 'allowNoFastForward': False}}
```

### Suggested fix

Looking at the code in [policy.py](https://github.com/Azure/azure-devops-cli-extension/blob/5edcb95f88669cca898b7bd60f188328ca9c1804/azure-devops/azext_devops/dev/repos/policy.py#L297), we see this line:
`        # We cannot send setting as None in the API`

So it looks someone ran into this problem in the past and dealt with it correctly, but only when `is_new_merge_type_update` is true. The same fix needs to be done in the third part of this [if-elif-else chain](https://github.com/Azure/azure-devops-cli-extension/blob/5edcb95f88669cca898b7bd60f188328ca9c1804/azure-devops/azext_devops/dev/repos/policy.py#L314).

If you're open to me contributing, I can try fixing it myself. It's only a few lines, but if testing it will be a pain for the uninitiated I can leave it to you guys!

**Comment by gauravsaralMs** [comment]
@davidtorosyan : it will be great if you can contribute, it will also give us an idea how easy/ hard it is to contribute into this repo :)

**Comment by davidtorosyan** [comment]
I'm trying to get set up now, and I hit some issues:
1. The [dev_setup](https://github.com/Azure/azure-devops-cli-extension/blob/master/doc/dev_setup.md) docs were a little hard to find. I expected a link in [CONTRIBUTING.md](https://github.com/Azure/azure-devops-cli-extension/blob/master/CONTRIBUTING.md) but had to search around to find them instead.
2. The setup steps say to run `env\Scripts\activate.bat`, but this doesn't work in powershell and silently fails. I did find `activate.ps1` myself though.

Still hitting an issue though:

```powershell
> az extension list
Traceback (most recent call last):
  File "c:\users\user\appdata\local\programs\python\python38-32\Lib\runpy.py", line 192, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "c:\users\user\appdata\local\programs\python\python38-32\Lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\code\azure-devops-cli-extension\env\lib\site-packages\azure\cli\__main__.py", line 33, in <module>
    az_cli = get_default_cli()
  File "C:\code\azure-devops-cli-extension\env\lib\site-packages\azure\cli\core\__init__.py", line 546, in get_default_cli
    return AzCli(cli_name='az',
  File "C:\code\azure-devops-cli-extension\env\lib\site-packages\azure\cli\core\__init__.py", line 55, in __init__
    SESSION.load(os.path.join(azure_folder, 'az.sess'), max_age=3600)
  File "C:\code\azure-devops-cli-extension\env\lib\site-packages\azure\cli\core\_session.py", line 46, in load
    if st.st_mtime + max_age < time.clock():
AttributeError: module 'time' has no attribute 'clock'
```

This might be due to my Python version, but I'm not sure.
```powershell
> python --version
Python 3.8.0
```

**Comment by atbagga** [comment]
@davidtorosyan Probably yes. There is a known issue with python 3.8 which is why we also had to restrict our CI pipeline runs to 3.7.

**Comment by davidtorosyan** [comment]
I've got python reinstalled:

```powershell
> python --version
Python 3.7.5
```

And I successfully ran the setup:

```powershell
(env) PS C:\azure-devops-cli-extension> python scripts/dev_setup.py
Running dev setup...
Root directory 'C:\azure-devops-cli-extension'

Extension directory 'C:\azure-devops-cli-extension\azure-devops'

Azure root directory 'C:\Users\user\.azure'

Executing: python -m pip install --upgrade pip
Requirement already up-to-date: pip in c:\azure-devops-cli-extension\env\lib\site-packages (19.3.1)

Executing: pip install --pre azure-cli --extra-index-url https://azurecliprod.blob.core.windows.net/edge --no-cache-dir
Looking in indexes: https://pypi.org/simple, https://azurecliprod.blob.core.windows.net/edge
Requirement already satisfied: azure-cli in c:\azure-devops-cli-extension\env\lib\site-packages (2.0.76)
# ...

Executing: pip install -e C:\azure-devops-cli-extension\azure-devops
Obtaining file:///C:/azure-devops-cli-extension/azure-devops
Requirement already satisfied: distro==1.3.0 in c:\azure-devops-cli-extension\env\lib\site-packages (from azure-devops==0.14.0) (1.3.0)
# ...
Installing collected packages: azure-devops
  Found existing installation: azure-devops 0.14.0
    Uninstalling azure-devops-0.14.0:
      Successfully uninstalled azure-devops-0.14.0
  Running setup.py develop for azure-devops
Successfully installed azure-devops

Executing: pip install --upgrade --target C:\Users\user\.azure/devcliextensions/azure-devops C:\azure-devops-cli-extension\azure-devops
Processing c:\azure-devops-cli-extension\azure-devops
Collecting distro==1.3.0
  Using cached https://files.pythonhosted.org/packages/f6/b1/ba5a96bccd3496241d8908164b9502a129156443cdd5acbdbf04a90b7a09/distro-1.3.0-py2.py3-none-any.whl
# ...
Building wheels for collected packages: azure-devops
  Building wheel for azure-devops (setup.py) ... done
  Created wheel for azure-devops: filename=azure_devops-0.14.0-py2.py3-none-any.whl size=1405256 sha256=11a0772a351fddc7a425f76763df92a1d73472750768ef307ef2714d0713a6e7
  Stored in directory: C:\Users\user\AppData\Local\pip\Cache\wheels\fe\a1\fa\b262e971bd298654c51defdbbf48ca202855adcb74d6c6d5b1
Successfully built azure-devops
Installing collected packages: distro, six, python-dateutil, certifi, urllib3, idna, chardet, requests, oauthlib, requests-oauthlib, isodate, msrest, azure-devops
Successfully installed azure-devops-0.14.0 certifi-2019.9.11 chardet-3.0.4 distro-1.3.0 idna-2.8 isodate-0.6.0 msrest-0.6.10 oauthlib-3.1.0 python-dateutil-2.7.3 requests-2.22.0 requests-oauthlib-1.2.0 six-1.13.0 urllib3-1.25.6

Finished dev setup.
```

But I'm not having any luck after that:

```powershell
> az extension list
[]

> az devops -h
az: 'devops' is not in the 'az' command group. See 'az --help'.
```

**Comment by atbagga** [comment]
@davidtorosyan Did you try these from within the virtual environment? 
```
az extension list
az devops -h
```
I tried the dev setup instructions on a fresh windows VM and it seems to work for me. I will look into the dev setup instructions to make it better. 

For temporary workaround you can build the extension wheel with the changes and install it locally to test it, if you want.

## Build a wheel 
In azure-devops-cli-extension/azure-devops directory run -
```
python setup.py bdist_wheel
```

A new extension wheel (azure_devops-0.X.0-py2.py3-none-any.whl) will be generated in the dist directory (azure-devops-cli-extension\azure-devops\dist). 

# Install this wheel 
Install from local source with command -
```
az extension add --source <PATH_TO_WHL_FILE>
```

Test your command changes.

**Comment by atbagga** [comment]
@davidtorosyan Are you still in to give a try to your setup or should I go ahead and fix it?

**Comment by davidtorosyan** [comment]
I don’t think I’ll have time to set up the temporary workaround. I did run those commands you mentioned in the virtual environment - those were the logs I posted above. You can go ahead with the fix, I’ll just keep an eye out for other ways I can contribute!

**Comment by atbagga** [comment]
Thank you for interest in the contribution. We mark the issues with "help wanted" tag for contributors to pick up. 
I will try to repro your dev setup scenario and we will revisit the dev setup instructions too. 

I will raise a fix for this issue.

**Comment by atbagga** [comment]
Fixed in version 0.16.0. 
ETA for release end of this week.
